### PR TITLE
Implicit link paths

### LINKED ISSUE #7115: Improve Spack's detection of compiler run paths
Spack compiler detection does a pretty good job of finding compiler executables, but there are still configuration issues with certain vendor compilers.  Specifically, users have to add `extra_rpaths` to their compiler entries in `compilers.yaml`, especially for vendor compilers and newer `gcc`/`clang` where the `libstdc++` version is not the same as the system.

Spack should be able to autodetect the necessary compiler runtime paths so that it can add correct `RPATHs` to executables.  This task will add this feature.

- [x] Improve compiler detection so that each compiler can tell Spack how to detect its runtime paths.  Basic functionality can be implemented generically in the `Compiler` class and specialized when needed. (#7153)

- [x] Update the Spack build (`build_environment.py`) so that RPATHs for compiler runtime paths are added automatically via Spack's compiler wrappers. (#7411)

- [x] show that intel compilers no longer need manually configured `extra_rpaths`

### PR DESCRIPTION
**Pull request by mathstuf**
Resolves #7115.

Currently Spack assumes that compilers add `RPATH`s for any non-standard runtime directories they use.  For example, versions of `g++` and the Intel compiler that support newer versions of C++ need an updated `libstdc++.so`, so they link against a `libstdc++.so` in `$gcc_prefix/lib64`.  We call `$gcc_prefix/lib64` here an "implicit link path" because the compiler implicitly links against libaries there that were not included on the command line.

While Spack-built compilers add `RPATH`s for these directories, compilers installed at many sites do not.  This leads to the following type of error when you run binaries built with these compilers:

```
/tmp/spack00t/spack-stage/cmake-3.15.1-ywecvmgyzvulocevd5dmilq55zsz5jlp/spack-src/Bootstrap.cmk/cmake: /lib64/libstdc++.s
o.6: version `GLIBCXX_3.4.20' not found (required by /tmp/spack00t/spack-stage/cmake-3.15.1-ywecvmgyzvulocevd5dmilq55zsz5
jlp/spack-src/Bootstrap.cmk/cmake)
```

The solution so far has been to manually add `extra_rpaths` to `compilers.yaml`, which forces Spack to add the `RPATH`s, or to tell users to load the compiler's module when running with binaries it generated.  Neither of these is ideal.

This PR eliminates those requirements by doing a few things:
- [x] Spack now auto-detects implicit link paths when it detects compilers, by querying compilers for what their implicit link paths are.  This is borrowed from [what CMake does](https://cmake.org/cmake/help/v3.0/variable/CMAKE_LANG_IMPLICIT_LINK_LIBRARIES.html) to detect implicit link paths.
- [x] Spack adds these paths to `compilers.yaml` in a new `implicit_rpaths` field:

    ```yaml
    compilers:
    - compiler:
        spec:intel@19.0.1
        paths:
          cc: /opt/intel-19.0.1/bin/icc
          cxx: /opt/intel-19.0.1/bin/icpc
          f77: /opt/intel-19.0.1/bin/ifort
          fc: /opt/intel-19.0.1/bin/ifort
        environment: {}
        # spack now detects these implicit paths automatically and sets them here
        implicit_rpaths:
        - /opt/intel-19.0.1/compilers_and_libraries_2019.4.243/linux/compiler/lib/intel64_lin
        - /opt/gcc-4.9.3/lib64            
        # you can still edit extra_rpaths but likely do not need to
        extra_rpaths: []
        flags: {}
        modules: []
        operating_system: rhel7
        target: x86_64
    ```
- [x] Spack's compiler wrappers now automatically inject implicit RPATHs into binaries they build.

So, in summary, you likely no longer need to fiddle with `extra_rpaths`.  If you do have to, file an issue, and we'll see if we need to add more detection for it.

### CODE REVIEW

File: lib/spack/spack/compiler.py [hunk 1]
```diff
@@ -174,6 +176,105 @@ def check(exe):
...
+    # a verbose flag is available
+    @property
+    def verbose_flag(self):
+        # If it is not overridden, assume it is not supported
+        pass
+
+    _LINKER_LINE = \
+        re.compile(r'^( *|.*[/\\])'
+                   r'(link|ld|([^/\\]+-)?ld|collect2)'
+                   r'[^/\\]*( |$)')
```
Comment (mathstuf) [inline lib/spack/spack/compiler.py]: This looks dumb because `flake8` is picky.

**Comment by aharrison24** [comment]
This will be really useful!

With this branch checked out on an ubuntu 16.04 VM, I installed llvm inside a spack instance and did this:
```
$ spack load llvm
$ spack compiler find
```

The clang compiler was picked up correctly, and the implicit link paths were deduced as:
```
    implicit_link_paths:
    - /opt/spack_llvm_bootstrap/opt/spack/linux-ubuntu16.04-x86_64/gcc-5.4.0/llvm-5.0.1-a7juikjjoqn3fvldllpyohi44tumybrn/lib
    - /usr/lib/x86_64-linux-gnu
    - /usr/lib/gcc/x86_64-linux-gnu/5.4.0
    - /usr/lib
    - /lib64
    - /lib
    - /lib/x86_64-linux-gnu
```

Is it correct/expected that it would be picking up those library paths related to the system's gcc compiler?

**Comment by mathstuf** [comment]
> Is it correct/expected that it would be picking up those library paths related to the system's gcc compiler?

That depends on the build. It may have compiled those in during its build.

**Comment by aharrison24** [comment]
OK. So this could be an issue with the `llvm` package in spack?

For what it's worth, here is the output of a verbose build using the clang++ compiler from the llvm package.
https://gist.github.com/aharrison24/d4448432d06c2fdc74a54a74a686e40b
In the `ld` line at the bottom, all of those system library paths are listed before the spack llvm library paths.

So it does seem like the PR is doing the right thing in this case and the behaviour I'm seeing is caused by issues elsewhere.

Will there be a separate PR to have spack act upon the information in the `implicit_link_paths` field of `compilers.yaml`? Can't find any reference to that anywhere.

**Comment by mathstuf** [comment]
> Will there be a separate PR to have spack act upon the information in the implicit_link_paths field of compilers.yaml? Can't find any reference to that anywhere.

Yes, that is the rest of #7115 and #7120.

File: lib/spack/spack/compiler.py (hunk 1 above)
Comment (tgamblin) [reply lib/spack/spack/compiler.py]: I usually avoid the `\` by breaking after the paren, then indent the contents by 4.

Comment (mathstuf) [reply lib/spack/spack/compiler.py]: Yeah, that still requires a line break though since the `re.compile(…)` hits 80 with the extra indentation.

File: lib/spack/spack/compiler.py [hunk 2]
```diff
@@ -174,6 +176,103 @@ def check(exe):
     def version(self):
         return self.spec.version
 
+    # This property should be overridden in the compiler subclass if
+    # a verbose flag is available
```
Comment (tgamblin) [inline lib/spack/spack/compiler.py]: make into docstring?

File: lib/spack/spack/compiler.py [hunk 3]
```diff
@@ -174,6 +176,103 @@ def check(exe):
...
+            tmpdir = tempfile.mkdtemp(prefix='spack-implicit-link-info')
+            fout = os.path.join(tmpdir, 'output')
+            fin = os.path.join(tmpdir, 'main.c')
+
+            with open(fin, 'w+') as csource:
+                csource.write(
+                    'int main(int argc, char* argv[]) { '
+                    '(void)argc; (void)argv; return 0; }\n')
+                csource.flush()
+            output = Executable(first_compiler)(self.verbose_flag, fin, '-o',
```
Comment (tgamblin) [inline lib/spack/spack/compiler.py]: can `Executable(first_compiler)` be named something to make this a bit cleaner?

File: lib/spack/spack/compiler.py [hunk 4]
```diff
@@ -174,6 +176,103 @@ def check(exe):
...
+                        continue
+                    else:
+                        lib_search_paths = False
+                elif line.startswith('Library search paths:'):
+                    lib_search_paths = True
+
+                if not self._LINKER_LINE.match(line):
+                    pass
+                if self._LINKER_LINE_IGNORE.match(line):
+                    pass
```
Comment (tgamblin) [inline lib/spack/spack/compiler.py]: same question

File: lib/spack/spack/compiler.py [hunk 5]
```diff
@@ -174,6 +176,103 @@ def check(exe):
...
+                        continue
+
+                    if next_arg:
+                        raw_link_dirs.append(arg)
+                        next_arg = False
+                        continue
+
+                    if self._LINK_DIR_ARG.match(arg):
+                        tty.debug('linkdir: %s' % arg)
+                        raw_link_dirs.append(arg[2:])
```
Comment (tgamblin) [inline lib/spack/spack/compiler.py]: instead of remembering the index here and on the next condition, can you extract the linkdir and libpath with parens?  That will be cleaner if the expressions change.

File: lib/spack/spack/compiler.py [hunk 6]
```diff
@@ -174,6 +176,103 @@ def check(exe):
...
+                    elif self._LIBPATH_ARG.match(arg):
+                        tty.debug('libpath: %s', arg)
+                        raw_link_dirs.append(arg[9:])
```
Comment (tgamblin) [inline lib/spack/spack/compiler.py]: see above.

File: lib/spack/spack/compiler.py [hunk 7]
```diff
@@ -174,6 +176,103 @@ def check(exe):
...
+            tty.debug('found link dirs: %s' % ', '.join(implicit_link_dirs))
+            return implicit_link_dirs
+        finally:
+            shutil.rmtree(tmpdir, ignore_errors=True)
+
+    __implicit_link_paths = None
+
+    @property
+    def implicit_link_paths(self):
+        if self.__implicit_link_paths is not None:
```
Comment (tgamblin) [inline lib/spack/spack/compiler.py]: Why two underscores?  Is that intentional (ala https://shahriar.svbtle.com/underscores-in-python?). Seems like just one would be sufficient.

Comment (mathstuf) [reply lib/spack/spack/compiler.py]: I meant it to be private. Subclasses shouldn't care about the caching here; if they really want to they should override the `implicit_link_paths` property instead.

**Comment by mathstuf** [comment]
Updated for the requested changes. Unit tests are still missing. Though, I'm not sure what to expect in those tests.

**Comment by tgamblin** [comment]
@mathstuf: can you just make the tests simple scripts that exercise the parsing logic?  Basically I would just run each compiler with the `verbose_flag`, and make sure that for the output, the expected implicit paths are generated.

**Comment by mathstuf** [comment]
> expected implicit paths are generated.

This is what I don't know to write in the tests. A compiler loaded from `/opt/intel-16.10` is going to differ from one in `/opt/intel-17.10`. Should I just expect a non-empty list?

**Comment by tgamblin** [comment]
> This is what I don't know to write in the tests. A compiler loaded from /opt/intel-16.10 is going to differ from one in /opt/intel-17.10. Should I just expect a non-empty list?

Maybe I was vague.  I mean you could run a compiler on the cluster and take its output and put *that* in the test suite to exercise things.  So the test suite shouldn't actually interact with a compiler -- just some saved output, for which you'd know what to expect.

**Comment by mathstuf** [comment]
Well, that'd make the test pass (or not-skipped) on that one machine and fail (or skip) on everything else. I doubt the efficacy of such a test.

**Comment by tgamblin** [comment]
It's not supposed to be a machine-specific test, just a spot test to prevent regression in the parser.  Let's get the output of several different compilers (on several machines if you want), and make sure `_determine_implicit_link_paths()` works on it.  The test would be something like:

```
lib/spack/spack/test/data/
    implicit_link_paths/
        icc-16.0.1.out
        icc-17.2.3.out
        xl-10.1.out
        clang-3.9.0.out
```

And the test would read those in and validate that `implicit_link_paths()` works on at least some inputs we know about.  If we add code changes, we can avoid some regressions.

Agreed that it's not going to validate every machine, but it's better than nothing.

**Comment by mathstuf** [comment]
That still doesn't help. My `clang-3.9.0` is in `/usr`. On Apple, it's under some Homebrew directory. Someone else might have it as a module under `/opt/clang/3.9.0`. There's no general answer to "oh, you have this compiler, I know the paths you're supposed to be using". It's like trying to test `os.path.expanduser('~')` with `/home/boeckb` as the expected value; it doesn't work as a unit test.

File: lib/spack/spack/compiler.py [hunk 8]
```diff
@@ -174,6 +176,111 @@ def check(exe):
     def version(self):
         return self.spec.version
 
+    @property
+    def verbose_flag(self):
+        """
+        This property should be overridden in the compiler subclass if a
+        verbose flag is available.
+
+        If it is not overridden, it is assumed to not be supported.
```
Comment (tgamblin) [inline lib/spack/spack/compiler.py]: This isn't actually true -- if this is not overridden, it looks like `_determine_implicit_link_paths` still uses the result.  I think it should just skip detection.

Comment (mathstuf) [reply lib/spack/spack/compiler.py]: In `implicit_link_paths`, the cache is set if the flag isn't set otherwise `_determine_implicit_link_paths` is called to fill the cache.

**Comment by tgamblin** [comment]
> That still doesn't help. My clang-3.9.0 is in /usr. On Apple, it's under some Homebrew directory. Someone else might have it as a module under /opt/clang/3.9.0. There's no general answer to "oh, you have this compiler, I know the paths you're supposed to be using". It's like trying to test os.path.expanduser('~') with /home/boeckb as the expected value; it doesn't work as a unit test.

I added two simple tests to show what I'm talking about.  Again, it's not a system test, it's a parser regression test.  All I had to do was factor out the parsing part.  Can you add some more for other compilers?

There's another question in here which is where we should filter system paths.  We're going to `RPATH` these things, so system paths shouldn't be added to the `RPATH`.  There is `spack.util.environment.filter_system_paths()` that we'll probably need in here.

**Comment by mathstuf** [comment]
Ah, I see now. Sorry, I missed that you wanted a refactor to separate out the inner logic. This makes more sense now.

I updated it to also check that we don't extract incorrect paths from the compilers.

> There's another question in here which is where we should filter system paths. We're going to RPATH these things, so system paths shouldn't be added to the RPATH. There is spack.util.environment.filter_system_paths() that we'll probably need in here.

I think it should be filtered where RPATHs are handled. The implicit link paths are used no matter what. Spack should filter where it adds rpaths, not in the storage of the implicit link paths.

**Comment by mathstuf** [comment]
Updated with Clang and GCC outputs from my machine. None of the LLNL machines I saw had XL or Cray compilers. I'll see if I can get output from around here.

**Comment by mathstuf** [comment]
Updated with xlc and cce example outputs.

**Comment by mathstuf** [comment]
Cc: @scheibelp

**Comment by mathstuf** [comment]
Implementing #7411 got me to see that the following was missing here:

  - updating the json schema for `compilers.yaml`
  - filling the `implicit_link_paths` from `compilers.yaml`

Now fixed.

**Comment by scheibelp** [comment]
> I am also not sure about the distinction between extra_rpaths and implicit_link_paths in compilers.yaml. Should they be the same? Do we need two attributes for this? Currently, spack.build_environment already adds stuff in extra_rpaths to the RPATHs passed to the compiler wrappers.

@tgamblin the separate extra_rpaths/implicit_link_paths is currently used to determine whether the compiler has already checked for implicit link paths. If the entries were not separate, and if the user were to for example create their own compilers.yaml entry with extra_rpaths already defined, then Spack would not know to add the implicit paths. Using the current strategy, compiler.extra_rpaths could still be the only thing that the API user needs to worry about - the `Compiler` object can grab `implicit_link_paths` and add that under the covers.

The overall strategy could be changed: Spack-discovered compilers (either with `spack compiler find`, or `spack compiler add...`) could add the implicit paths at initialization time, and for entries manually created by a user there could be a `spack compiler add-implicit-links-paths` command or something like that. Overall the current approach would be more convenient for the user (since they wouldnt need to worry about running an extra command for manual entries).

This could also be avoided by creating a scope that is maintained entirely by spack (and it maintains the extra_rpaths entry in that scope). I prefer maintaining the separate config entries to this though because the scope precedence interactions may be confusing (and from the implementation perspective I think maintaining separate entries is about the same amount of work and is also clearer).

**Comment by scheibelp** [comment]
To summarize https://github.com/spack/spack/pull/7153#issuecomment-375073923, it isn't necessary to maintain separate config entries, but I don't think it will make things harder for the user if we do, and I don't think it will be harder to maintain.

**Comment by scheibelp** [comment]
@tgamblin: this is an update on the current state of the PR. I'm currently considering a few ways to achieve the appropriate balance in filtering paths.

Originally this was failing because the following implicit link directories are detected for gcc 4.8 (i.e. this is from `~/.spack/linux/compilers.yaml`):

```
implicit_rpaths:
- /usr/lib/gcc/x86_64-linux-gnu/4.8
- /usr/lib/x86_64-linux-gnu
- /lib/x86_64-linux-gnu
```

In particular, `/usr/lib/x86_64-linux-gnu` contains `libpython2.7.so`; when installing python, the make target instructions point `LD_LIBRARY_PATH` to the `libpython2.7.so` created for that python installation, but this is overridden by the RPATH (which takes precedence).

Currently this is addressed by following up after `spack.util.environment.filter_system_paths` by additionally removing paths which contain `/usr/` or `/usr/lib/` *as a prefix* whereas `filter_system_paths` only removes paths if they exactly match.

`filter_system_paths` should keep its original behavior because it includes `/usr/local/` and AFAIK that could be managed using per-package prefixes (which would not be auto-included by `ld`). Arguably this also means that the current compiler link path filtering is too aggressive.

I can think of three ways to proceed with this:

* check candidate directories for the presence of compiler libraries (e.g. `libgcc.a`) - this isn't necessarily foolproof since they could appear in a shared prefix
* check whether the path explicitly mentions a compiler: this is an even looser heuristic but would work here
* The most-aggressive but safest option I can think of is to specifically look for a spack metadata directory relative to the listed binaries. One of the main goals is to avoid RPATH errors with spack-built compilers and this would address that directly without above problems with including too much.

File: lib/spack/env/cc [hunk 9]
```diff
@@ -426,6 +426,12 @@ case "$mode" in
             rpaths+=("${extra_rpaths[@]}")
         fi
 
+        # Set implicit RPATHs
+        IFS=':' read -ra implicit_rpaths <<< "$SPACK_COMPILER_IMPLICIT_RPATHS"
+        if [[ "$add_rpaths" != "false" ]] ; then
+            rpaths+=("${implicit_rpaths[@]}")
+        fi
```
Comment (tgamblin) [inline lib/spack/env/cc]: Rather than adding a redundant section in `cc`, I think it's easier if the compiler wrappers just have one list of RPATHs, and we deal with the detail that there are two kinds inside `build_environment.py`.  Is there any reason to do the same thing twice here when we could just add these to `SPACK_COMPILER_EXTRA_RPATHS`?

File: lib/spack/spack/build_environment.py [hunk 10]
```diff
@@ -348,6 +348,10 @@ def set_build_environment_variables(pkg, env, dirty):
         extra_rpaths = ':'.join(compiler.extra_rpaths)
         env.set('SPACK_COMPILER_EXTRA_RPATHS', extra_rpaths)
 
+    if compiler.implicit_rpaths:
+        implicit_rpaths = ':'.join(compiler.implicit_rpaths)
+        env.set('SPACK_COMPILER_IMPLICIT_RPATHS', implicit_rpaths)
```
Comment (tgamblin) [inline lib/spack/spack/build_environment.py]: same question: can we just appent to `extra_rpaths` here?

File: lib/spack/spack/compiler.py [hunk 11]
```diff
@@ -152,6 +251,53 @@ def check(exe):
...
 
+    @classmethod
+    def verbose_flag(cls):
+        """
+        This property should be overridden in the compiler subclass if a
+        verbose flag is available.
+
+        If it is not overridden, it is assumed to not be supported.
+        """
+        pass
```
Comment (tgamblin) [inline lib/spack/spack/compiler.py]: no need for `pass` if the function has a docstring.

File: lib/spack/spack/compiler.py [hunk 12]
```diff
@@ -152,6 +251,53 @@ def check(exe):
...
+
+    @classmethod
+    def parse_implicit_rpaths(cls, string):
```
Comment (tgamblin) [inline lib/spack/spack/compiler.py]: Just remove this classmethod and make `_parse_implicit_rpaths` non-private.

File: lib/spack/spack/test/data/compilers.yaml [hunk 13]
```diff
@@ -8,6 +8,7 @@ compilers:
       f77: None
       fc: None
     modules: 'None'
+    implicit_rpaths: []
```
Comment (tgamblin) [inline lib/spack/spack/test/data/compilers.yaml]: Is this required now?  If so we should avoid that, as it'll break existing configs that do not have this field.  We should also do some separate PR that makes the YAML for this stuff more minimal by omitting empty entries.

File: share/spack/qa/run-build-tests [hunk 14]
```diff
@@ -41,4 +41,4 @@ fi
 spack config get compilers
 
 # Run some build smoke tests, potentially with code coverage
-${coverage_run} bin/spack install ${OPTIONS} ${SPEC}
+${coverage_run} bin/spack -d install ${OPTIONS} ${SPEC}
```
Comment (tgamblin) [inline share/spack/qa/run-build-tests]: do we want to make this change?  I think at the very least it doesn't belong in this PR.  I'm assuming that this was some vestigial change for debugging Travis.

File: lib/spack/spack/compiler.py (hunk 12 above)
Comment (becker33) [reply lib/spack/spack/compiler.py]: This is so that someone has a hook to add a compiler that we don't know about if its format isn't properly parsed by the overall method. Are you sure we want to get rid of that?

File: lib/spack/env/cc (hunk 9 above)
Comment (becker33) [reply lib/spack/env/cc]: This ensures that the user-added paths in `extra_rpaths` are added before the generated ones in `implicit_rpaths`. I'm not sure whether its important though.

File: lib/spack/spack/compiler.py (hunk 12 above)
Comment (tgamblin) [reply lib/spack/spack/compiler.py]: good point.  No -- let's leave it.

File: lib/spack/spack/test/data/compilers.yaml (hunk 13 above)
Comment (becker33) [reply lib/spack/spack/test/data/compilers.yaml]: This PR does not brack existing configs that do not have this field. Confirmed on my mac.

File: lib/spack/env/cc (hunk 9 above)
Comment (tgamblin) [reply lib/spack/env/cc]: Yeah my point is that the build_environment can take care of that, since extra_rpaths are ordered.

File: lib/spack/spack/test/data/compilers.yaml (hunk 13 above)
Comment (tgamblin) [reply lib/spack/spack/test/data/compilers.yaml]: ok good.  I don't think we need to add this here then do we?

Comment (tgamblin) [reply lib/spack/spack/test/data/compilers.yaml]: e.g., we already do not have an `extra_rpaths` field -- why change it?

File: lib/spack/env/cc (hunk 9 above)
Comment (becker33) [reply lib/spack/env/cc]: I got halfway through doing this and realized we actually handle them differently. We generate `-L` flags for `extra_rpaths` and not for `implicit_rpaths`.

Comment (tgamblin) [reply lib/spack/env/cc]: so we should still put them in the order you're saying, I'm just saying it doesn't require additional code in `cc`.

Comment (becker33) [reply lib/spack/env/cc]: It does require additional code in `cc`. The directories from `extra_rpaths` are put in both `libdirs` and `rpaths`, but the directories from `implicit_rpaths` are only put in `rpaths`.

Comment (tgamblin) [reply lib/spack/env/cc]: Ok this is a good reason to keep them separate. Thanks.

**Comment by adamjstewart** [comment]
This looks great! The only question I had seems to have already been asked: why are we creating an extra field called `implicit_rpaths` instead of auto-populating `extra_rpaths`? I'll assume most compilers have been thoroughly tested. Let me know if you need/want any additional compilers or old versions to be tested.

**Comment by becker33** [comment]
> why are we creating an extra field called `implicit_rpaths` instead of auto-populating `extra_rpaths`

@adamjstewart because the `extra_rpaths` field could still be useful to add libraries into the build. I'm not 100% sure, but I think we could use it to swap openmp runtimes, for example. 

Libraries in `extra_rpaths` have `-L` and `-Wl,-rpath` flags generated. The libraries in `implicit_rpaths` have implied `-L` flags, so we only need to generate the `-Wl,-rpath` flags.

**Comment by tgamblin** [comment]
The other main gist of it is that `extra_rpaths` is for people and `implicit_rpaths` is for Spack.  If we had one field, Spack could clobber something someone wrote.  We wanted to avoid that.

**Comment by adamjstewart** [comment]
@becker33 getting more reports of erroneous implicit RPATHs causing build failures: https://github.com/spack/spack/issues/12709#issuecomment-529501935
